File: c/gpt/p121_nr_ver2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(void) {
    const size_t chunk_size = 10; /* bytes */
    char *chunk1 = (char *)calloc(chunk_size, 1);
    char *chunk2 = (char *)calloc(chunk_size, 1);

    if (chunk1 == NULL || chunk2 == NULL) {
        free(chunk1);
        free(chunk2);
        return 1;
    }

    const char *src = "123456789"; /* 9 chars + '\0' fits in 10 bytes */
    memcpy(chunk1, src, strlen(src) + 1);

    size_t to_copy = strnlen(chunk1, chunk_size - 1);
    memcpy(chunk2, chunk1, to_copy);
    chunk2[to_copy] = '\0';

    printf("%s\n", chunk2);

    free(chunk1);
    chunk1 = NULL;

    printf("%s\n", chunk2);

    free(chunk2);
    chunk2 = NULL;

    return 0;
}

File: java/gpt/p121_nr_ver2.java
import java.nio.charset.StandardCharsets;

public class p121_nr_ver2 {
    private static String toStringFromNullTerminated(byte[] buf) {
        int end = 0;
        while (end < buf.length && buf[end] != 0) end++;
        return new String(buf, 0, end, StandardCharsets.UTF_8);
    }

    public static void main(String[] args) {
        final int chunkSize = 10;
        byte[] chunk1 = new byte[chunkSize];
        byte[] chunk2 = new byte[chunkSize];

        byte[] src = "123456789".getBytes(StandardCharsets.UTF_8); // 9 bytes
        int n = Math.min(src.length, chunkSize - 1);
        System.arraycopy(src, 0, chunk1, 0, n);
        chunk1[n] = 0;

        int n2 = Math.min(n, chunkSize - 1);
        System.arraycopy(chunk1, 0, chunk2, 0, n2);
        chunk2[n2] = 0;

        System.out.println(toStringFromNullTerminated(chunk2));

        chunk1 = null;

        System.out.println(toStringFromNullTerminated(chunk2));

        chunk2 = null;
    }
}